ggml/src/ggml-vulkan/vulkan-shaders/vulkan-shaders-gen.cpp

```diff
@@ -232,28 +233,105 @@ std::string basename(const std::string &path) {
     return path.substr(path.find_last_of("/\\") + 1);
 }
 
+std::stringstream make_generic_stringstream() {
+    std::stringstream ss;
+    ss.imbue(c_locale);
+    return ss;
+}
+
+std::vector<unsigned char> read_binary_file(const std::string& path, bool may_not_exist = false) {
+    FILE* f = fopen(path.c_str(), "rb");
+    if (!f) {
+        if (!may_not_exist) {
+            std::cerr << "Error opening file: " << path << " (" << strerror(errno) << ")\n";
+        }
+        return {};
+    }
+
+    fseek(f, 0, SEEK_END);
+    size_t size = ftell(f);
+    fseek(f, 0, SEEK_SET);
+
+    std::vector<unsigned char> data(size);
+    size_t read_size = fread(data.data(), 1, size, f);
+    fclose(f);
+    if (read_size != size) {
+        std::cerr << "Error reading file: " << path << " (" << strerror(errno) << ")\n";
+        return {};
+    }
+
+    return data;
+}
+
+void write_binary_file(const std::string& path, const unsigned char * data, size_t size) {
+    FILE* f = fopen(path.c_str(), "wb");
+    if (!f) {
+        std::cerr << "Error opening file for writing: " << path << " (" << strerror(errno) << ")\n";
+        return;
+    }
+
+    size_t write_size = fwrite(data, 1, size, f);
+    fclose(f);
+    if (write_size != size) {
+        std::cerr << "Error writing file: " << path << " (" << strerror(errno) << ")\n";
+        return;
+    }
+}
+
+void write_file_if_changed(const std::string& path, const std::string& content) {
+    std::vector<unsigned char> existing = read_binary_file(path, true);
+    if (existing.size() != content.size() || memcmp(existing.data(), content.data(), content.size()) != 0) {
```

Comment:
I think this is preventing dependencies from being satisfied. If I touch a shader source but don't change it, it'll keep recompiling it each time I build.

Reply:
Yes, at least for generators which check output file timestamps.

I've changed the .cpp files to be written unconditionally. They compile quickly now that they're much smaller.

For the .hpp I think the check is still a good trade-off - it is quick and often avoids compilation of ggml-vulkan.cpp which is very slow.